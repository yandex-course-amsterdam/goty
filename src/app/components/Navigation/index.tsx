import React, { ReactElement } from 'react'

import style from './style.css'

type NavigationProps = {
  children?: React.ReactNode
  title?: string
}

export const Navigation = ({
  children,
  title
}: NavigationProps): ReactElement => {
  return (
    <nav className={style.navigation}>
      {title && <p className={style.title}>{title}</p>}
      {children}
    </nav>
  )
}
