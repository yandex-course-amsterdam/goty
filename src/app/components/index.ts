export { Sidebar } from './Sidebar'
export { Avatar } from './Avatar'
export { Navigation } from './Navigation'
export { Main } from './Main'
export { Title } from './Title'
export { Description } from './Description'
export { Input } from './Input'
export { Textarea } from './Textarea'
export { Label } from './Label'
export { Error } from './Error'
export { ErrorBoundary } from './ErrorBoundary'
export { OfflineBar } from './OfflineBar'
export { Button } from './Button'
export { SubNavigation } from './SubNavigation'
export { Canvas } from './Canvas'
export { UserScore } from './UserScore'
export { Comment } from './Comment'
export { Comments } from './Comments'
export { Likes } from './Likes'
export { NewsList } from './NewsList'
export { SignUpForm } from './Forms/SignUpForm'
export { SignInForm } from './Forms/SignInForm'
export { FeedbackForm } from './Forms/FeedbackForm'
export { DetailsForm } from './Forms/DetailsForm'
export { PasswordForm } from './Forms/PasswordForm'
export { AvatarForm } from './Forms/AvatarForm'
export { Loader } from './Loader'
export { PrivateRoute } from './PrivateRoute'
export { ExitButton } from './ExitButton'
export { Authorization } from './Authorization'
export { Theme } from './Theme'
export { App } from './App'
