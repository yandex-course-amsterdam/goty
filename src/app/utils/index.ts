export { sanitize } from './sanitize'
export { displayResponseText } from './displayResponseText'
export { startServiceWorker } from './startServiceWorker'
export * from './storage'
