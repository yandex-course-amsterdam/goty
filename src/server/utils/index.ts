export { createToken } from './createToken'
export { verifyToken } from './verifyToken'
