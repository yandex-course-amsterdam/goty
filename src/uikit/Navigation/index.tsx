export { Container } from './Container'
export { ListItem } from './Item'
export { List } from './List'
