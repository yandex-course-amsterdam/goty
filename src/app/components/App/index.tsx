import React, { FC, useCallback, useEffect, useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'

import { Switch, Route } from 'react-router-dom'
import { CSSTransition } from 'react-transition-group'

import {
  SignUpView,
  SignInView,
  ProfileView,
  GameView,
  ScoreView,
  NotFoundView
} from 'app/views'
import { OfflineBar, PrivateRoute, Authorization } from 'app/components'

import { postResult, getUserTheme } from 'app/api/Api'
import { fetchUserInfo } from 'app/actions'

import { StoreState } from 'app/reducers'
import { route } from 'app/enums'
import {
  getScore,
  removeScore,
  isServer,
  setUserTheme,
  storeUserTheme
} from 'app/utils'

import 'normalize.css'
import 'assets/main.css'
import '../../../fonts/fonts.css'

// startServiceWorker()

export const App: FC = (): JSX.Element => {
  const [isOffline, setIsOffline] = useState(isServer || !navigator.onLine)

  const dispatch = useDispatch()

  const loginStatus = useSelector(
    (state: StoreState) => state.loginStatus.status
  )

  const handleOfflineCallback = useCallback(() => setIsOffline(true), [])
  const handleOnlineCallback = useCallback(() => {
    setIsOffline(false)

    const storedScore = getScore()

    if (storedScore) {
      postResult(storedScore)
      removeScore()
    }
  }, [])

  useEffect(() => {
    const fetchUser = async () => {
      try {
        const data = await dispatch(fetchUserInfo())
        // Очередной странный тайпинг, учитывая возвращаемое по факту значение
        // Сейчас уже нет времени разбираться, буду потом сильнее вникать в ts
        // @ts-ignore
        const { data: theme } = await getUserTheme(data.id)
        setUserTheme(theme)
        storeUserTheme(theme)
      } catch (error) {
        console.warn(error)
      }
    }

    if (loginStatus) {
      fetchUser()
    }

    if (!isServer) {
      window.addEventListener('offline', handleOfflineCallback)
      window.addEventListener('online', handleOnlineCallback)
    }

    return () => {
      if (!isServer) {
        window.removeEventListener('offline', handleOfflineCallback)
        window.removeEventListener('online', handleOnlineCallback)
      }
    }
  }, [])

  return (
    <>
      <Switch>
        <Route exact path={route.auth}>
          <Authorization />
        </Route>
        <Route path={route.signUp}>
          <SignUpView />
        </Route>
        <Route path={route.signIn}>
          <SignInView />
        </Route>
        <PrivateRoute path={route.profile}>
          <ProfileView />
        </PrivateRoute>
        <PrivateRoute path={route.game}>
          <GameView />
        </PrivateRoute>
        <PrivateRoute path={route.score}>
          <ScoreView />
        </PrivateRoute>
        <Route path="*">
          <NotFoundView />
        </Route>
      </Switch>

      <CSSTransition
        in={isOffline}
        appear={isOffline}
        timeout={0}
        classNames="bar"
      >
        <OfflineBar />
      </CSSTransition>
    </>
  )
}
