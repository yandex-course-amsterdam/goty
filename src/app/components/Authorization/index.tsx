import React, { FC, useEffect, useState } from 'react'
import { useDispatch } from 'react-redux'
import { useLocation, Redirect } from 'react-router-dom'

import { authWithYandexOauth } from 'app/api/Api'
import { fetchUserInfo, setOAuthStatus } from 'app/actions'
import { LoaderView } from 'app/views'
import { route } from 'app/enums'

const useQuery = () => new URLSearchParams(useLocation().search)

export const Authorization: FC = (): JSX.Element => {
  const [isLoading, setIsLoading] = useState(true)
  const dispatch = useDispatch()

  const query = useQuery()
  const codeString = query.get('code')

  const getUserData = async () => {
    try {
      await dispatch(fetchUserInfo())
      setIsLoading(!isLoading)
    } catch (error) {
      console.error(error)
      setIsLoading(!isLoading)
    }
  }

  const signInWithOAuth = async () => {
    try {
      await authWithYandexOauth(codeString)
      await dispatch(fetchUserInfo())
      dispatch(setOAuthStatus(true))
      setIsLoading(!isLoading)
    } catch (error) {
      console.log(error)
      setIsLoading(!isLoading)
    }
  }

  useEffect(() => {
    if (codeString) {
      signInWithOAuth()
    } else {
      getUserData()
    }
  }, [])

  return isLoading ? <LoaderView /> : <Redirect to={route.game} />
}
