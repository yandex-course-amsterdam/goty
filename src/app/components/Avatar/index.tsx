import React, { FC } from 'react'
import { useSelector } from 'react-redux'
import cn from 'classnames'

import { StoreState } from 'app/reducers'

import style from './style.css'

interface IProps {
  className?: string
}

export const Avatar: FC<IProps> = ({ className }): JSX.Element => {
  const userInfo = useSelector((state: StoreState) => state.userInfo)

  const userAvatar = userInfo.avatar
    ? `https://ya-praktikum.tech/${userInfo.avatar}`
    : '/images/avatar.png'

  const avatarName = userInfo.displayName || 'Top game'

  return (
    <div className={cn(style.avatar, className)}>
      <img className={style.image} src={userAvatar} alt={avatarName} />
      <h1 className={style.title}>{avatarName}</h1>
    </div>
  )
}
