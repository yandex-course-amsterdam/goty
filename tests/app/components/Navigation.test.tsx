import React from 'react'
import * as renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { createMemoryHistory } from 'history'

import { Navigation } from 'app/components/Navigation'

import { renderWithRouter } from '../../utils'

interface IUserInfo {
  login: string | null
}

const initialState: Record<string, IUserInfo> = {
  userInfo: {
    login: 'Oh hi'
  }
}
const mockStore = configureStore()
const store = mockStore(initialState)

describe('Navigation component', () => {
  test('Render itself correctly', () => {
    const history = createMemoryHistory()
    const component = renderer
      .create(
        renderWithRouter(
          <Provider store={store}>
            <Navigation />
          </Provider>,
          history
        )
      )
      .toJSON()
    expect(component).toMatchSnapshot()
  })
})
