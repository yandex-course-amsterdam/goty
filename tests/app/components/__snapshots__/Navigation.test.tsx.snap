// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component Render itself correctly 1`] = `
<nav>
  <p>
    Options
  </p>
  <ul
    className=""
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <li
        className=""
      >
        <svg
          fill="none"
          height="32"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 27.428c6.312 0 11.428-5.116 11.428-11.428C27.428 9.688 22.312 4.57 16 4.57 9.688 4.571 4.57 9.688 4.57 16S9.688 27.428 16 27.428z"
            stroke="#787885"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
          <path
            d="M16 20.572a4.571 4.571 0 100-9.143 4.571 4.571 0 000 9.143zM7.92 7.92l4.846 4.846M19.234 19.234l4.846 4.846M19.234 12.766L24.08 7.92M19.234 12.766L23.27 8.73M7.92 24.08l4.846-4.846"
            stroke="#787885"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
        </svg>
        <p>
          Game
        </p>
      </li>
    </a>
    <a
      href="/profile/details"
      onClick={[Function]}
    >
      <li
        className=""
      >
        <svg
          fill="none"
          height="32"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M25.143 26.286V24a4.57 4.57 0 00-4.572-4.571H11.43A4.571 4.571 0 006.857 24v2.286M16 14.857a4.571 4.571 0 100-9.143 4.571 4.571 0 000 9.143z"
            stroke="#787885"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
        </svg>
        <p>
          Profile
        </p>
      </li>
    </a>
    <a
      href="/score/leaderboard"
      onClick={[Function]}
    >
      <li
        className=""
      >
        <svg
          fill="none"
          height="32"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 19.429a8 8 0 100-16 8 8 0 000 16z"
            stroke="#787885"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
          <path
            d="M11.668 18.16L10.287 28.57 16 25.143l5.714 3.428-1.383-10.423"
            stroke="#787885"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
        </svg>
        <p>
          Score
        </p>
      </li>
    </a>
  </ul>
  <button
    className=""
    onClick={[Function]}
    type="button"
  >
    <li
      className=""
    >
      <svg
        fill="none"
        height="56"
        width="56"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="#2A2731"
          height="56"
          rx="28"
          width="56"
        />
        <path
          d="M27.5 22.929L21.93 28.5l5.571 5.572M33.072 28.5H21.929"
          stroke="#787885"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="3"
        />
      </svg>
      <p>
        Exit
      </p>
    </li>
  </button>
</nav>
`;
